### PR TITLE
SONARJS-627 CFG and UnreachableCode

### PR DESCRIPTION
**Pull request by pynicolas**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+    }
+  }
+
+  private void build(List<? extends Tree> trees) {
+    for (Tree tree : Lists.reverse(trees)) {
+      build(tree);
+    }
+  }
+
+  private void build(Tree tree) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: MAJOR The Cyclomatic Complexity of this method "build" is 21 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/BranchingBlock.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  @Override
+  public Set<MutableBlock> successors() {
+    Preconditions.checkState(successor1 != null, "Successors were not set on " + this);
+    return ImmutableSet.of(successor1, successor2);
+  }
+
+  public void setSuccessors(MutableBlock successor1, MutableBlock successor2) {
+    Preconditions.checkArgument(successor1 != null && successor2 != null, "Successor cannot be null");
+    Preconditions.checkArgument(successor1 != this && successor2 != this, "Cannot add itself as successor");
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/BranchingBlock.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)
MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/BranchingBlock.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    this.successor1 = successor1;
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/BranchingBlock.java]: BLOCKER NullPointerException might be thrown as 'this' is nullable here [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+      immutableBlockByMutable.put(mutableBlock, immutableBlock);
+      immutableBlockSetBuilder.add(immutableBlock);
+      index++;
+    }
+    immutableBlockByMutable.put(end, this.end);
+    
+    ImmutableSetMultimap.Builder<ControlFlowNode, ControlFlowNode> successorBuilder = ImmutableSetMultimap.builder();
+    ImmutableSetMultimap.Builder<ControlFlowNode, ControlFlowNode> predecessorBuilder = ImmutableSetMultimap.builder();
+    ImmutableSetMultimap.Builder<ControlFlowNode, SyntaxToken> jumpBuilder = ImmutableSetMultimap.builder();
+    for (MutableBlock mutableBlock : blocks) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java]: MAJOR Combine this loop with the one that starts on line 65. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3047)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java [hunk 5]
```diff
@@ -50,107 +51,61 @@
...
     }
   }
 
-  private void enterBlock() {
-    blockLevel.push(false);
-  }
-
-  private void exitBlock() {
-    blockLevel.pop();
+  private Tree unreachableTree(List<Tree> elements) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java]: MAJOR Make "unreachableTree" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* BLOCKER 1 blocker
* MAJOR 5 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-javascript/blob/e51ce3a1c7111d27d930efe3e4bc24d24765ab5a/javascript-frontend/src/main/java/org/sonar/javascript/cfg/EndBlock.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/SingleSuccessorBlock.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+  public abstract MutableBlock successor();
+
+  @Override
+  public Set<MutableBlock> successors() {
+    Preconditions.checkState(successor() != null, "No successor was set on " + this);
+    return ImmutableSet.of(successor());
+  }
+
+  public MutableBlock firstNonEmptySuccessor() {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/SingleSuccessorBlock.java]: @pynicolas So this method may return empty successor if it's not `instanceof SingleSuccessorBlock`, mayn't it?

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/SingleSuccessorBlock.java]: @pynicolas And it could return `this` if it's empty. It conflicts with name

Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/SingleSuccessorBlock.java]: @vilchik-elena This method is also used when removing empty blocks. BranchingBlocks cannot be empty,  the EndBlock is empty by nature, but is not removed and has no successor. I will think of a better name for this method...

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/SingleSuccessorBlock.java]: @pynicolas may be `skipEmptyBlocks` and add comment line that "BranchingBlocks cannot be empty, the EndBlock is empty by nature"?

Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/SingleSuccessorBlock.java]: @vilchik-elena done in 2394dbaf3174a819307e64cec7adb2410d8b835c

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-javascript/blob/2394dbaf3174a819307e64cec7adb2410d8b835c/javascript-frontend/src/main/java/org/sonar/javascript/cfg/EndBlock.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-javascript/blob/edbc11e4ce2f6dd84c25a6711e377d28cd72a650/javascript-frontend/src/main/java/org/sonar/javascript/cfg/EndBlock.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+      realConditionBlock.setSuccessors(loopBodyBlock, forStatementSuccessor);
+    }
+
+    if (fakeConditionBlock != null) {
+      fakeConditionBlock.setSuccessor(loopBodyBlock);
+    }
+
+    if (tree.init() == null) {
+      currentBlock = createSimpleBlock(conditionBlock);
+      if (tree.condition() == null && loopBodyBlock.isEmpty()) {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @pynicolas Could explain this piece?

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+    MutableBlock target = null;
+    String label = tree.label() == null ? null : tree.label().name();
+    for (Breakable breakable : breakables) {
+      if (label == null || label.equals(breakable.label)) {
+        target = breakable.breakTarget;
+        break;
+      }
+    }
+    Preconditions.checkState(target != null, "No break target can be found for label " + label);
+    currentBlock.addDisconnectingJump(tree.breakKeyword());
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @pynicolas In my understanding we should `target.addDisconnectingJump` as this method adds tokens with jump to this block. What is wrong with my reasoning?

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java (hunk 7 above)
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @vilchik-elena We have to handle the very unlikely case of an infinite loop with no statement: `for(;;){}`. In this case, we need at least one "block" which is not empty (otherwise, calling skipEmptyBlocks never ends...). That's why create a block with the `for` keyword, which is a bit artificial. After removing the empty block, this new block will have only one successor: itself.

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java (hunk 8 above)
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @vilchik-elena as discussed, it's `currentBlock` which is disconnected and that's the information we need to implement `UnreachableCodeCheck`.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-javascript/blob/c366ec1fb29312e455f6d5dc2ac6f7104005595e/javascript-frontend/src/main/java/org/sonar/javascript/cfg/EndBlock.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+      // Nothing to do
+    } else {
+      throw new IllegalArgumentException("Cannot build CFG for " + tree);
+    }
+  }
+
+  private void visitBlock(BlockTree block) {
+    boolean hasLabel = currentLabel != null;
+    if (hasLabel) {
+      addBreakable(null);
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @pynicolas Could you explain why we add breakable only for block and loops?

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+      currentBlock = createSimpleBlock(currentBlock);
+    }
+
+    build(block.statements());
+
+    if (hasLabel) {
+      breakables.removeFirst();
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @pynicolas better to have method `removeBreakable` (as there is `addBreakable`)

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java (hunk 9 above)
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @vilchik-elena And `switch`. When else should we add a breakable? `label1: var x;` does not raise any syntax error, but it does not create any label.

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java (hunk 10 above)
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @vilchik-elena done in 15c62c0

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java (hunk 9 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @pynicolas Ok, indeed, looks like there is no other way to get working label

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-javascript/blob/15c62c0c9cc7deb24db194a26beb9c8e9e17950a/javascript-frontend/src/main/java/org/sonar/javascript/cfg/EndBlock.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+    build(tree.statement());
+  }
+
+  private void visitTryStatement(TryStatementTree tree) {
+    MutableBlock catchOrFinallyBlock = null;
+
+    if (tree.finallyBlock() != null) {
+      currentBlock = createSimpleBlock(currentBlock);
+      build(tree.finallyBlock());
+      throwTargets.push(currentBlock);
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @pynicolas So if we `throw` inside `catch` block we will go to `finally` block?

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+      }
+    }
+
+    blocks.removeAll(emptyBlockReplacements.keySet());
+
+    for (MutableBlock block : blocks) {
+      block.replaceSuccessors(emptyBlockReplacements);
+    }
+
+    if (emptyBlockReplacements.containsKey(start)) {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @pynicolas IMO `start` field is unnecessary, we can use `currentBlock` instead.

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java (hunk 11 above)
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @vilchik-elena yes

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+  ControlFlowGraph(Set<MutableBlock> blocks, MutableBlock start, MutableBlock end) {
+
+    Map<MutableBlock, ControlFlowNode> immutableBlockByMutable = new HashMap<>();
+    ImmutableSet.Builder<ControlFlowBlock> immutableBlockSetBuilder = ImmutableSet.builder();
+    int index = 0;
+    for (MutableBlock mutableBlock : blocks) {
+      ImmutableBlock immutableBlock = new ImmutableBlock(index, mutableBlock.elements());
+      immutableBlockByMutable.put(mutableBlock, immutableBlock);
+      immutableBlockSetBuilder.add(immutableBlock);
+      index++;
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java]: @pynicolas I don't understand the point of this index if you are iterating over set (it should be unstable).

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java (hunk 12 above)
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @vilchik-elena Having a `start` field is maybe not strictly required, but improves understandability. `currentBlock` changes too often.

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+    }
+    immutableBlockByMutable.put(end, this.end);
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java]: @pynicolas Why? You should have already done this in previous loop

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java]: @pynicolas Ah, you replacing in the map ImmutableBlock by EndNode.

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java (hunk 13 above)
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java]: @vilchik-elena Its only purpose is to give different and readable ids to the graph blocks. We will probably need to have a stable order in the future to have stable results in symbolic execution.

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java (hunk 14 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java]: @pynicolas No, I don't understand: `end` should be in `blocks` but there is no in fact.

Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java]: @vilchik-elena No, `end` is not supposed to be part of `blocks`. The end node is special: it does not implement `ControlFlowBlock` since it cannot have any element.

File: javascript-frontend/src/test/java/org/sonar/javascript/cfg/MutableBlockTest.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+public class MutableBlockTest {
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  private final Tree tree1 = mock(Tree.class);
+  private final EndBlock end = new EndBlock();
+  private final SimpleBlock simple1 = new SimpleBlock(end);
+  private final SimpleBlock simple2 = new SimpleBlock(end);
+  private final BranchingBlock branching1 = new BranchingBlock(tree1);
```
Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/cfg/MutableBlockTest.java]: @pynicolas rename to `branching`? (there is no second)
the same for `tree1`

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/BranchingBlock.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+  @Override
+  public Set<MutableBlock> successors() {
+    Preconditions.checkState(successor1 != null, "Successors were not set on " + this);
+    return ImmutableSet.of(successor1, successor2);
+  }
+
+  public void setSuccessors(MutableBlock successor1, MutableBlock successor2) {
+    Preconditions.checkArgument(successor1 != null && successor2 != null, "Successor cannot be null");
+    Preconditions.checkArgument(!this.equals(successor1) && !this.equals(successor2), "Cannot add itself as successor");
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/BranchingBlock.java]: @pynicolas Why did you add this assertion?

File: javascript-frontend/src/test/java/org/sonar/javascript/cfg/MutableBlockTest.java (hunk 15 above)
Comment (pynicolas) [inline javascript-frontend/src/test/java/org/sonar/javascript/cfg/MutableBlockTest.java]: @vilchik-elena as you prefer

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/BranchingBlock.java (hunk 16 above)
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/BranchingBlock.java]: @vilchik-elena To catch stupid mistakes as early as possible: the way we build the CFG, a branching block cannot have itself as a successor.

File: javascript-frontend/src/test/java/org/sonar/javascript/cfg/ControlFlowGraphTest.java [hunk 17]
```diff
@@ -0,0 +1,640 @@
...
+  public void no_block() throws Exception {
+    ControlFlowGraph g = build("", 0);
+    assertThat(g.start()).isEqualTo(g.end());
+  }
+
+  @Test
+  public void single_basic_block() throws Exception {
+    ControlFlowGraph g = build("foo();", 1);
+    assertThat(g.start()).isEqualTo(g.blocks().iterator().next());
+    assertThat(g.start().predecessors()).isEmpty();
```
Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/cfg/ControlFlowGraphTest.java]: @pynicolas  Can't we move this asserting into `build`?

Comment (pynicolas) [inline javascript-frontend/src/test/java/org/sonar/javascript/cfg/ControlFlowGraphTest.java]: @vilchik-elena I don't think so. This assertion is specific to this test case.

Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/cfg/ControlFlowGraphTest.java]: @pynicolas Ah, indeed.

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java (hunk 13 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraph.java]: @pynicolas And it's not used now (I thought you use them in tests, but there are different indexes there). May be we better remove them to not be confused?

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-javascript/blob/319fd0a9cbb27523582ee2d9cb4872b0744080b3/javascript-frontend/src/main/java/org/sonar/javascript/cfg/EndBlock.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java [hunk 18]
```diff
@@ -50,107 +51,61 @@
...
-
-    } else if (tree.is(Kind.IF_STATEMENT)) {
-      return !((IfStatementTree) tree).statement().is(Kind.BLOCK);
-
-    } else if (tree.is(Kind.ELSE_CLAUSE)) {
-      return !((ElseClauseTree) tree).statement().is(Kind.BLOCK);
-
+    if (tree.is(Kind.SCRIPT)) {
+      check(ControlFlowGraph.build((ScriptTree) tree));
+    } else if (tree.is(Kind.METHOD, Kind.GENERATOR_METHOD)) {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java]: @pynicolas method trees also extend FunctionTree

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-javascript/blob/a7c98c96ad53dc8b8d39b2edeb9ae6aec0b94ef0/javascript-frontend/src/main/java/org/sonar/javascript/cfg/EndBlock.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)
